lib/Parser/Parse.cpp

```diff
@@ -13815,8 +13827,84 @@ void ParseNode::Dump()
         break;
     }
 }
+
+void DumpCapturedNames(ParseNodeFnc* pnodeFnc, IdentPtrSet* capturedNames, ArenaAllocator* alloc)
+{
+    auto sortedNames = JsUtil::List<IdentPtr, ArenaAllocator>::New(alloc);
+
+    capturedNames->Map([=](const IdentPtr& pid) -> void {
+        sortedNames->Add(pid);
+    });
+
+    sortedNames->Sort([](void* context, const void* left, const void* right) -> int {
+        const IdentPtr leftIdentPtr = *(const IdentPtr*)(left);
+        const IdentPtr rightIdentPtr = *(const IdentPtr*)(right);
+        return ::wcscmp(leftIdentPtr->Psz(), rightIdentPtr->Psz());
+    }, nullptr);
+
+    sortedNames->Map([=](int index, const IdentPtr pid) -> void {
+        OUTPUT_TRACE_DEBUGONLY(Js::CreateParserStatePhase, _u("Function %u captured name \"%s\"\n"), pnodeFnc->functionId, pid->Psz());
+    });
+}
 #endif
 
+void Parser::AddNestedCapturedNames(ParseNodeFnc* pnodeChildFnc)
+{
+    if (m_currentNodeFunc && this->IsCreatingStateCache() && pnodeChildFnc->HasAnyCapturedNames())
+    {
+        IdentPtrSet* parentCapturedNames = GetCurrentFunctionNode()->EnsureCapturedNames(&m_nodeAllocator);
+        IdentPtrSet* childCaptureNames = pnodeChildFnc->GetCapturedNames();
+
+        auto iter = childCaptureNames->GetIterator();
+
+        while (iter.IsValid())
+        {
+            parentCapturedNames->AddNew(iter.CurrentValue());
+            iter.MoveNext();
+        }
+    }
+}
+
+void Parser::ProcessCapturedNames(ParseNodeFnc* pnodeFnc)
+{
+    if (this->IsCreatingStateCache() && pnodeFnc->HasAnyCapturedNames())
+    {
+        IdentPtrSet* capturedNames = pnodeFnc->GetCapturedNames();
+        auto iter = capturedNames->GetIteratorWithRemovalSupport();
+
+        while (iter.IsValid())
+        {
+            const IdentPtr& pid = iter.CurrentValueReference();
+            PidRefStack* ref = pid->GetTopRef();
+
+            // If the pid has no refs left in our function's scope after binding, we didn't capture it.
+            if (!ref || ref->GetScopeId() < pnodeFnc->pnodeBodyScope->blockId)
+            {
+                iter.RemoveCurrent();
+            }
+
+            iter.MoveNext();
+        }
+
+#if DBG_DUMP
+        if (Js::Configuration::Global.flags.Trace.IsEnabled(Js::CreateParserStatePhase))
+        {
+            DumpCapturedNames(pnodeFnc, capturedNames, &this->m_nodeAllocator);
+            fflush(stdout);
+        }
+#endif
+    }
+}
+
+bool Parser::IsCreatingStateCache()
+{
+    return this->m_parseType == ParseType_StateCache
```

Comment:
>this->m_parseType == ParseType_StateCache [](start = 11, length = 41)

curious what makes this true?

Reply:
At the moment, nothing, which is why there is a force config flag. This flag would be pushed down from the API entrypoint which kicks-off the parse, like ParseScriptText